doc/source/reference/c-api/array.rst

```diff
@@ -2884,11 +3172,279 @@ Data-type descriptors
+                PyArray_DTypeMeta *dtype1, PyArray_DTypeMeta *dtype2)
+
+   Given two input DTypes, determines the appropriate "common" DType
+   that can store values for both types. Returns ``Py_NotImplemented``
+   if no such type exists.
+
+.. c:macro:: NPY_DT_common_instance
+
+.. c:type:: PyArray_Descr *(PyArrayDTypeMeta_CommonInstance)( \
+               PyArray_Descr *dtype1, PyArray_Descr *dtype2)
+
+   Given two input descriptors, determines the appropriate "common"
+   descriptor that can store values for both instances. Returns ``NULL``
+   on error.
+
+.. c:macro:: NPY_DT_ensure_canonical
+
+.. c:type:: PyArray_Descr *(PyArrayDTypeMeta_EnsureCanonical)( \
+                PyArray_Descr *dtype)
+
+   Returns the "canonical" representation for a descriptor instance. The
+   notion of a canonical descriptor generalizes the concept of byte
+   order, in that a canonical descriptor always has native byte
+   order. If the descriptor is already canonical, this function returns
+   a new reference to the input descriptor.
+
+.. c:macro:: NPY_DT_setitem
+
+.. c:type:: int(PyArrayDTypeMeta_SetItem)(PyArray_Descr *, PyObject *, char *)
+
+   Implements scalar setitem for an array element given a PyObject.
+
+.. c:macro:: NPY_DT_getitem
+
+.. c:type:: PyObject *(PyArrayDTypeMeta_GetItem)(PyArray_Descr *, char *)
+
+   Implements scalar getitem for an array element. Must return a python
+   scalar.
+
+.. c:macro:: NPY_DT_get_clear_loop
+
+   If defined, sets a traversal loop that clears data in the array. This
+   is most useful for arrays of references that must clean up array
+   entries before the array is garbage collected. Implements
+   ``PyArrayMethod_GetLoop``.
+
+.. c:macro:: NPY_DT_get_fill_zero_loop
+
+   If defined, sets a traversal loop that fills an array with "zero"
+   values, which may have a DType-specific meaning. This is called
+   inside `numpy.zeros` for arrays that need to write a custom sentinel
+   value that represents zero if for some reason a zero-filled array is
+   not sufficient. Implements ``PyArrayMethod_GetLoop``.
+
+.. c:macro:: NPY_DT_finalize_descr
+
+.. c:type:: PyArray_Descr *(PyArrayDTypeMeta_FinalizeDescriptor)( \
+                PyArray_Descr *dtype)
+
+   If defined, a function that is called to "finalize" a descriptor
+   instance after an array is created. One use of this function is to
+   force newly created arrays to have a newly created descriptor
+   instance, no matter what input descriptor is provided by a user.
+
+PyArray_ArrFuncs slots
+^^^^^^^^^^^^^^^^^^^^^^
+
+In addition the above slots, the following slots are exposed to allow
+filling the ``PyArray_ArrFuncs`` struct attached to descriptor
+instances.
+
+.. c:macro:: NPY_DT_PyArray_ArrFuncs_getitem
+
+   Allows setting a per-dtype getitem. Note that this is not necessary
+   to define unless the default version calling the function defined
+   with the ``NPY_DT_getitem`` ID is unsuitable for some reason.
```

Comment:
Why would the general function be unsuitable?

Reply:
Some places might call getitem in an arguably invalid way (passing `NULL` as the array).  I would hope NumPy got rid of all of them and things are fine, but I guess you might run into it somewhere.  Plus, optimize it.

(Not sure how much it makes sense to get into the weeds of it?)

Reply:
I added some clarifying text incorporating sebastian's explanation.